File: backend/app/schemas/sales_order.py
"""销售单 schemas + 红线脱敏构造工厂。转销售(锁档报价→销售单)本增量:只读投影,无写 schema(create 走 convert)。

🔴红线:客户售价(unit_price)/ 行额(line_total)/ 成交额(total_amount)对无 `receivable:read`
者置 null。可见性下沉到 `.build(..., can_see_price=...)` 构造工厂(单点经 redact_fields),
让「新出口忘记脱敏而裸露售价」在结构上更难发生,而非 dump 后再 patch。
SALES / FINANCE 持 receivable:read;PURCHASER / LOGISTICS 不持(浏览 SO 发起采购/出库,不看对客售价)。
"""
from __future__ import annotations

from datetime import datetime

from pydantic import BaseModel, Field

from app.rbac.redaction import SO_LINE_PRICE_FIELDS, SO_PRICE_FIELDS, redact_fields


class SalesOrderCancelIn(BaseModel):
    """整单取消:原因留痕(可空,上限挡误粘贴)。"""
    reason: str | None = Field(default=None, max_length=500)


class SalesOrderOut(BaseModel):
    id: int
    no: str
    source_quotation_id: int
    customer_id: int
    salesperson_id: int
    language: str
    currency: str
    status: str
    total_amount: float | None    # 🔴 无 receivable:read → null
    summary: str | None
    remark: str | None
    cancelled_at: datetime | None
    cancel_reason: str | None
    updated_at: datetime

    @classmethod
    def build(cls, so, extra: dict | None = None, *, can_see_price: bool) -> dict:
        d = cls.model_validate(so, from_attributes=True).model_dump()
        d = redact_fields(d, SO_PRICE_FIELDS, visible=can_see_price)
        if extra:
            d.update(extra)
        return d


class SalesOrderLineOut(BaseModel):
    id: int
    sales_order_id: int
    sku_id: int
    source_quotation_line_id: int
    name_snapshot: str
    spec_text_snapshot: str
    unit_snapshot: str
    unit_price: float | None      # 🔴 无 receivable:read → null
    qty: float
    line_total: float | None      # 🔴 无 receivable:read → null
    remark: str | None
    language: str
    sort_order: int

    @classmethod
    def build(cls, line, extra: dict | None = None, *, can_see_price: bool) -> dict:
        d = cls.model_validate(line, from_attributes=True).model_dump()
        d = redact_fields(d, SO_LINE_PRICE_FIELDS, visible=can_see_price)
        if extra:
            d.update(extra)
        return d


class SalesOrderListItem(BaseModel):
    id: int
    no: str
    summary: str | None
    customer_display: str
    salesperson_display: str
    status: str
    currency: str
    total_amount: float | None    # 🔴 无 receivable:read → null
    line_count: int
    created_at: datetime
    # 采购进度(轴2 派生,非红线):NOT_ORDERED/PARTIALLY_ORDERED/FULLY_ORDERED。
    purchase_progress: str | None = None

    @classmethod
    def build(cls, item, *, can_see_price: bool) -> dict:
        d = cls.model_validate(item).model_dump()
        return redact_fields(d, SO_PRICE_FIELDS, visible=can_see_price)

File: frontend/src/lib/salesOrder.ts
// 销售单前端类型 + API。对齐后端 schemas/sales_order.py。创建走报价 convert;写面 = 整单取消。
import { api } from "./api";
import type { Page } from "./catalog";
import { formatMoney } from "./format";
import type { PurchaseProgress, RelatedPurchaseOrder } from "./purchaseOrder";
import type { StockBalanceLine } from "./inventory";
import type { RelatedOutboundOrder } from "./outboundOrder";
import { qs } from "./qs";

// 镜像后端 SalesOrderStatus:CANCELLED 终态(报价回锁档可重转);更多态留给后续增量。
export type SalesOrderStatus = "CONFIRMED" | "CANCELLED";

/** 销售单行(读)。快照平移自报价行,冻结不变;source_quotation_line_id 记来源报价行。 */
export interface SalesOrderLineOut {
  id: number;
  sales_order_id: number;
  sku_id: number;
  source_quotation_line_id: number;
  name_snapshot: string;
  spec_text_snapshot: string;
  unit_snapshot: string;
  // 🔴 客户售价红线:无 receivable:read 者后端置 null(PURCHASER / LOGISTICS)。渲染走 formatPrice。
  unit_price: number | string | null;
  qty: number | string;
  line_total: number | string | null;
  remark: string | null;
  language: string;
  sort_order: number;
  // 采购增量扩展:该行已被采购覆盖的数量(非红线,始终可见)。剩余 = qty − covered_qty。
  covered_qty?: number | string;
}

export interface SalesOrderOut {
  id: number;
  no: string;
  source_quotation_id: number;
  customer_id: number;
  salesperson_id: number;
  language: string;
  currency: string;
  status: SalesOrderStatus;
  total_amount: number | string | null;  // 🔴 同上,无 receivable:read → null
  summary: string | null;
  remark: string | null;
  cancelled_at: string | null;
  cancel_reason: string | null;
  updated_at: string;
  // 详情响应附带(GET /{id} 服务端直出),create 响应不含,故可选。
  customer_display?: string;
  salesperson_display?: string;
  source_quotation_no?: string | null;
  // 采购增量扩展:采购进度(详情/列表均附带)。
  purchase_progress?: PurchaseProgress;
  // 关联采购单:仅 purchase:read 者的详情响应附带;SALES 无此字段(不渲染该区块)。
  related_purchase_orders?: RelatedPurchaseOrder[];
  // 库存增量扩展:仅持 inventory:read 者的详情响应附带该键(契约 §3 条件下发);
  // 前端不自判权限,响应键存在与否驱动渲染。
  stock_balances?: StockBalanceLine[];
  // 出库增量扩展:仅持 outbound:read 者的详情响应附带该键(条件下发,同上模式)。
  related_outbound_orders?: RelatedOutboundOrder[];
}

export interface SalesOrderListItem {
  id: number;
  no: string;
  summary: string | null;
  customer_display: string;
  salesperson_display: string;
  status: SalesOrderStatus;
  currency: string;
  total_amount: number | string | null;  // 🔴 同上,无 receivable:read → null
  line_count: number;
  created_at: string;
  // 采购增量扩展:采购进度徽标(列表附带)。
  purchase_progress?: PurchaseProgress;
}

export interface SalesOrderCancelBlockingDocument {
  type: "purchase_order" | "outbound_order";
  id: number;
  no: string;
  status: string;
  path: string;
}

export interface SalesOrderCancelBlockedData {
  blocking_kind?: "purchase_order" | "outbound_order";
  next_action?: string;
  blocking_documents?: SalesOrderCancelBlockingDocument[];
}

/** 🔴 售价红线渲染:null(无 receivable:read)→ 静音「—」,绝不显示 0。镜像采购侧 formatCost。 */
export function formatPrice(v: number | string | null | undefined): string {
  return v === null || v === undefined ? "—" : formatMoney(v);
}

export interface SalesOrderListFilters {
  status?: string;
  customer_id?: number;
  salesperson_id?: number;
  no?: string;
  purchase_progress?: string;
  // 采购台选单入口:只列可发起采购的 SO(排除已采完)。
  purchasable_only?: boolean;
  sort?: "created_at" | "total_amount";
  dir?: "asc" | "desc";
  page?: number;
  size?: number;
}

export const salesOrderApi = {
  list: (p: SalesOrderListFilters) =>
    api.get<Page<SalesOrderListItem>>(`/api/v1/sales-orders${qs(p as Record<string, unknown>)}`),
  get: (id: number) =>
    api.get<{ order: SalesOrderOut; lines: SalesOrderLineOut[] }>(`/api/v1/sales-orders/${id}`),
  // 整单取消(sales:manage):41802/41803 的 data 带真实下游阻塞单据列表。
  cancel: (id: number, reason?: string | null) =>
    api.post<{ order: SalesOrderOut; lines: SalesOrderLineOut[] }>(
      `/api/v1/sales-orders/${id}/cancel`, { reason: reason ?? null }),
};
